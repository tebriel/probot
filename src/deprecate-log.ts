import type { Logger } from "pino";
import { Deprecation } from "deprecation";

import type { DeprecatedLogger } from "./types";

/**
 * `app.log()` and `context.log()` were aliases for `.log.info()`.
 * The alias will be removed in v11
 */
export function deprecateLog(log: Logger): DeprecatedLogger {
  function logInfo() {
    log.warn(
      new Deprecation(
        '[probot] "app.log()" and "context.log()" are deprecated. Use "app.log.info()" and "context.log.info()" instead'
      )
    );
    // @ts-ignore
    log.info(...arguments);
  }

  for (const key in log) {
    // @ts-ignore
    logInfo[key] =
      typeof log[key] === "function" ? log[key].bind(log) : log[key];
  }

  // @ts-ignore
  return logInfo;
}
