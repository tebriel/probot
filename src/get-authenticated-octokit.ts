import { State } from "./types";

export async function getAuthenticatedOctokit(
  state: State,
  installationId?: number
) {
  const { githubToken, log, Octokit, octokit, throttleOptions } = state;

  if (!installationId) return octokit;

  const constructorAuthOptions = githubToken
    ? {}
    : { auth: { installationId: installationId } };
  const constructorThrottleOptions = throttleOptions
    ? {
        throttle: {
          id: installationId,
          ...throttleOptions,
        },
      }
    : {};

  const options = {
    log: log.child({ name: "github" }),
    ...constructorAuthOptions,
    ...constructorThrottleOptions,
  };

  return new Octokit(options);
}
