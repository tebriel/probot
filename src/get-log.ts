/**
 * A logger backed by [pino](https://getpino.io/)
 *
 * The default log level is `info`, but you can change it by setting the
 * `LOG_LEVEL` environment variable to `trace`, `debug`, `info`, `warn`,
 * `error`, or `fatal`.
 *
 * By default, logs are formatted for readability in development. If you intend
 * to drain logs to a logging service, set the `NODE_ENV` variable, e.g. `NODE_ENV=production probot run index.js`.
 *
 * ```js
 * app.log.debug("…so is this");
 * app.log.trace("Now we're talking");
 * app.log.info("I thought you should know…");
 * app.log.warn("Woah there");
 * app.log.error("ETOOMANYLOGS");
 * app.log.fatal("Goodbye, cruel world!");
 * ```
 */
import pino from "pino";
import type { LoggerOptions } from "pino";
import { getTransformStream } from "@probot/pino";

export function getLog() {
  const pinoOptions: LoggerOptions = {
    level: process.env.LOG_LEVEL || "info",
    name: "probot",
  };

  const transform = getTransformStream();
  transform.pipe(pino.destination(1));
  return pino(pinoOptions, transform);
}
