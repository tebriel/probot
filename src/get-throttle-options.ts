import Bottleneck from "bottleneck";
import Redis from "ioredis";

import { LoggerWithTarget } from "./wrap-logger";

type Options = {
  log: LoggerWithTarget;
  throttleOptions?: any;
  redisConfig?: Redis.RedisOptions;
};

export function getThrottleOptions(options: Options) {
  if (options.throttleOptions) {
    return options.throttleOptions;
  }

  if (options.redisConfig || process.env.REDIS_URL) {
    const connection = new Bottleneck.IORedisConnection({
      client: getRedisClient(options.redisConfig),
    });
    connection.on("error", options.log.error);

    return {
      Bottleneck,
      connection,
    };
  }
}

function getRedisClient(redisConfig?: Redis.RedisOptions): Redis.Redis | void {
  if (redisConfig) {
    return new Redis(redisConfig);
  }

  if (process.env.REDIS_URL) {
    return new Redis(process.env.REDIS_URL);
  }
}
