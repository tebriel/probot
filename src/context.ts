import path from "path";

import { Endpoints } from "@octokit/types";
import { EventNames, EventPayloads, WebhookEvent } from "@octokit/webhooks";
import merge from "deepmerge";
import yaml from "js-yaml";

import type { Logger } from "pino";

import { ProbotOctokit } from "./github/octokit";
import { deprecateLog } from "./deprecate-log";
import { DeprecatedLogger } from "./types";

type ReposGetContentsParams = Endpoints["GET /repos/:owner/:repo/contents/:path"]["parameters"];

const CONFIG_PATH = ".github";
const BASE_KEY = "_extends";
const BASE_REGEX = new RegExp(
  "^" +
  "(?:([a-z\\d](?:[a-z\\d]|-(?=[a-z\\d])){0,38})/)?" + // org
  "([-_.\\w\\d]+)" + // project
  "(?::([-_./\\w\\d]+\\.ya?ml))?" + // filename
    "$",
  "i"
);
const DEFAULT_BASE = ".github";

export type MergeOptions = merge.Options;

interface WebhookPayloadWithRepository {
  [key: string]: any;
  repository?: EventPayloads.PayloadRepository;
  issue?: {
    [key: string]: any;
    number: number;
    html_url?: string;
    body?: string;
  };
  pull_request?: {
    [key: string]: any;
    number: number;
    html_url?: string;
    body?: string;
  };
  sender?: {
    [key: string]: any;
    type: string;
  };
  action?: string;
  installation?: {
    id: number;
    [key: string]: any;
  };
}

/**
 * The context of the event that was triggered, including the payload and
 * helpers for extracting information can be passed to GitHub API calls.
 *
 *  ```js
 *  module.exports = app => {
 *    app.on('push', context => {
 *      context.log.info('Code was pushed to the repo, what should we do with it?');
 *    });
 *  };
 *  ```
 *
 * @property {github} github - A GitHub API client
 * @property {payload} payload - The webhook event payload
 * @property {logger} log - A logger
 */
export class Context<E extends WebhookPayloadWithRepository = any>
  implements WebhookEvent<E> {
  public name: EventNames.StringNames;
  public id: string;
  public payload: E;

  public github: InstanceType<typeof ProbotOctokit>;
  public log: DeprecatedLogger;

  constructor(
    event: WebhookEvent<E>,
    github: InstanceType<typeof ProbotOctokit>,
    log: Logger
  ) {
    this.name = event.name;
    this.id = event.id;
    this.payload = event.payload;

    this.github = github;
    this.log = deprecateLog(log);
  }

  // Maintain backward compatibility
  public get event(): string {
    return this.name;
  }

  /**
   * Return the `owner` and `repo` params for making API requests against a
   * repository.
   *
   * ```js
   * const params = context.repo({path: '.github/config.yml'})
   * // Returns: {owner: 'username', repo: 'reponame', path: '.github/config.yml'}
   * ```
   *
   * @param object - Params to be merged with the repo params.
   *
   */
  public repo<T>(object?: T) {
    const repo = this.payload.repository;

    if (!repo) {
      throw new Error(
        "context.repo() is not supported for this webhook event."
      );
    }

    return Object.assign(
      {
        owner: repo.owner.login || repo.owner.name!,
        repo: repo.name,
      },
      object
    );
  }

  /**
   * Return the `owner`, `repo`, and `issue_number` params for making API requests
   * against an issue. The object passed in will be merged with the repo params.
   *
   *
   * ```js
   * const params = context.issue({body: 'Hello World!'})
   * // Returns: {owner: 'username', repo: 'reponame', issue_number: 123, body: 'Hello World!'}
   * ```
   *
   * @param object - Params to be merged with the issue params.
   */
  public issue<T>(object?: T) {
    const payload = this.payload;
    return Object.assign(
      {
        issue_number: (payload.issue || payload.pull_request || payload).number,
      },
      this.repo(object)
    );
  }

  /**
   * Return the `owner`, `repo`, and `issue_number` params for making API requests
   * against an issue. The object passed in will be merged with the repo params.
   *
   *
   * ```js
   * const params = context.pullRequest({body: 'Hello World!'})
   * // Returns: {owner: 'username', repo: 'reponame', pull_number: 123, body: 'Hello World!'}
   * ```
   *
   * @param object - Params to be merged with the pull request params.
   */
  public pullRequest<T>(object?: T) {
    const payload = this.payload;
    return Object.assign(
      {
        pull_number: (payload.issue || payload.pull_request || payload).number,
      },
      this.repo(object)
    );
  }

  /**
   * Returns a boolean if the actor on the event was a bot.
   * @type {boolean}
   */
  get isBot() {
    return this.payload.sender!.type === "Bot";
  }

  /**
   * Reads the app configuration from the given YAML file in the `.github`
   * directory of the repository.
   *
   * For example, given a file named `.github/config.yml`:
   *
   * ```yml
   * close: true
   * comment: Check the specs on the rotary girder.
   * ```
   *
   * Your app can read that file from the target repository:
   *
   * ```js
   * // Load config from .github/config.yml in the repository
   * const config = await context.config('config.yml')
   *
   * if (config.close) {
   *   context.github.issues.comment(context.issue({body: config.comment}))
   *   context.github.issues.edit(context.issue({state: 'closed'}))
   * }
   * ```
   *
   * You can also use a `defaultConfig` object:
   *
   * ```js
   * // Load config from .github/config.yml in the repository and combine with default config
   * const config = await context.config('config.yml', {comment: 'Make sure to check all the specs.'})
   *
   * if (config.close) {
   *   context.github.issues.comment(context.issue({body: config.comment}));
   *   context.github.issues.edit(context.issue({state: 'closed'}))
   * }
   * ```
   *
   * Config files can also specify a base that they extend. `deepMergeOptions` can be used
   * to configure how the target config, extended base, and default configs are merged.
   *
   * For security reasons, configuration is only loaded from the repository's default branch,
   * changes made in pull requests from different branches or forks are ignored.
   *
   * @param fileName - Name of the YAML file in the `.github` directory
   * @param defaultConfig - An object of default config options
   * @param deepMergeOptions - Controls merging configs (from the [deepmerge](https://github.com/TehShrike/deepmerge) module)
   * @return Configuration object read from the file
   */
  public async config<T>(
    fileName: string,
    defaultConfig?: T,
    deepMergeOptions?: MergeOptions
  ): Promise<T | null> {
    const params = this.repo({ path: path.posix.join(CONFIG_PATH, fileName) });

    const config = await this.loadYaml(params);

    let baseRepo;
    if (config == null) {
      baseRepo = DEFAULT_BASE;
    } else if (config != null && BASE_KEY in config) {
      baseRepo = config[BASE_KEY];
      delete config[BASE_KEY];
    }

    let baseConfig;
    if (baseRepo) {
      if (typeof baseRepo !== "string") {
        throw new Error(`Invalid repository name in key "${BASE_KEY}"`);
      }

      const baseParams = this.getBaseParams(params, baseRepo);
      baseConfig = await this.loadYaml(baseParams);
    }

    if (config == null && baseConfig == null && !defaultConfig) {
      return null;
    }

    return (merge.all(
      // filter out null configs
      [defaultConfig, baseConfig, config].filter((conf) => conf),
      deepMergeOptions
    ) as unknown) as T;
  }

  /**
   * Loads a file from GitHub
   *
   * @param params Params to fetch the file with
   * @return The parsed YAML file
   */
  private async loadYaml<T>(params: ReposGetContentsParams): Promise<any> {
    try {
      // https://docs.github.com/en/rest/reference/repos#get-repository-content
      const response = await this.github.request(
        "GET /repos/{owner}/{repo}/contents/{path}",
        params
      );

      // Ignore in case path is a folder
      // - https://developer.github.com/v3/repos/contents/#response-if-content-is-a-directory
      if (Array.isArray(response.data)) {
        return null;
      }

      // we don't handle symlinks or submodule
      // - https://developer.github.com/v3/repos/contents/#response-if-content-is-a-symlink
      // - https://developer.github.com/v3/repos/contents/#response-if-content-is-a-submodule
      // tslint:disable-next-line
      if (typeof response.data.content !== "string") {
        return;
      }

      return (
        yaml.safeLoad(
          Buffer.from(response.data.content, "base64").toString()
        ) || {}
      );
    } catch (e) {
      if (e.status === 404) {
        return null;
      }

      throw e;
    }
  }

  /**
   * Computes parameters for the repository specified in base
   *
   * Base can either be the name of a repository in the same organization or
   * a full slug "organization/repo".
   *
   * @param params An object containing owner, repo and path
   * @param base A string specifying the base repository
   * @return The params of the base configuration
   */
  private getBaseParams(
    params: ReposGetContentsParams,
    base: string
  ): ReposGetContentsParams {
    const match = base.match(BASE_REGEX);
    if (match === null) {
      throw new Error(`Invalid repository name in key "${BASE_KEY}": ${base}`);
    }

    return {
      owner: match[1] || params.owner,
      path: match[3] || params.path,
      repo: match[2],
    };
  }
}
