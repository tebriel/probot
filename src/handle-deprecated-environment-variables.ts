import { getLog } from "./get-log";

export function handleDeprecatedEnvironmentVariables() {
  const log = getLog();

  // TODO: remove in v11
  if ("DISABLE_STATS" in process.env) {
    // tslint:disable:no-console
    log.warn('[probot] "DISABLE_STATS" has been removed in v10');
  }

  if ("IGNORED_ACCOUNTS" in process.env) {
    // tslint:disable:no-console
    log.warn('[probot] "IGNORED_ACCOUNTS" has been removed in v10');
  }

  if ("LOG_FORMAT" in process.env) {
    // tslint:disable:no-console
    log.warn(
      '[probot] "LOG_FORMAT" has been removed in v10. We now use https://getpino.io/ for logging. prettyPrint is enabled unless NODE_ENV is set to anything other than "development"'
    );
  }

  if ("LOG_LEVEL_IN_STRING" in process.env) {
    // tslint:disable:no-console
    log.warn(
      '[probot] "LOG_LEVEL_IN_STRING" has been removed in v10. Pipe the output to pino-text-level-transport instead. Example: `NODE_ENV=production probot run index.js | pino-text-level-transport`'
    );
  }

  if ("SENTRY_DSN" in process.env) {
    // tslint:disable:no-console
    log.warn(
      '[probot] "SENTRY_DSN" has been removed in v10. Use pino-sentry instead and make sure to set NODE_ENV. Example: `NODE_ENV=production probot run index.js | pino-sentry`'
    );
  }
}
