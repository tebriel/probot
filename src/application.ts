import { Webhooks } from "@octokit/webhooks";
import express from "express";
import Redis from "ioredis";
import LRUCache from "lru-cache";

import { ApplicationFunction } from ".";
import { Context } from "./context";
import { getAuthenticatedOctokit } from "./get-authenticated-octokit";
import { ProbotOctokit } from "./github/octokit";
import { logger } from "./logger";
import { LoggerWithTarget, wrapLogger } from "./wrap-logger";
import { getThrottleOptions } from "./get-throttle-options";
import { getProbotOctokitWithDefaults } from "./get-probot-octokit-with-defaults";
import { webhookTransform } from "./webhook-transform";
import { ProbotWebhooks, State } from "./types";
import { webhookEventCheck } from "./webhook-event-check";

export interface Options {
  // same options as Probot class
  privateKey?: string;
  githubToken?: string;
  id?: number;
  Octokit?: typeof ProbotOctokit;
  octokit?: InstanceType<typeof ProbotOctokit>;
  redisConfig?: Redis.RedisOptions;
  secret?: string;
  webhookPath?: string;

  // Application class specific options
  cache?: LRUCache<number, string>;
  throttleOptions?: any;
  webhooks?: Webhooks;

  // TODO: what is this for?
  router?: express.Router;
}

export type OnCallback<T> = (context: Context<T>) => Promise<void>;

/**
 * The `app` parameter available to `ApplicationFunction`s
 *
 * @property {logger} log - A logger
 */
export class Application {
  public router: express.Router;
  public log: LoggerWithTarget;
  public on: ProbotWebhooks["on"];
  public receive: ProbotWebhooks["receive"];

  private webhooks: ProbotWebhooks;
  private state: State;

  constructor(options: Options) {
    const opts = options;
    this.log = wrapLogger(logger);

    // TODO: support redis backend for access token cache if `options.redisConfig || process.env.REDIS_URL`
    const cache =
      options.cache ||
      new LRUCache<number, string>({
        // cache max. 15000 tokens, that will use less than 10mb memory
        max: 15000,
        // Cache for 1 minute less than GitHub expiry
        maxAge: Number(process.env.INSTALLATION_TOKEN_TTL) || 1000 * 60 * 59,
      });

    const Octokit = getProbotOctokitWithDefaults({
      githubToken: opts.githubToken,
      Octokit: opts.Octokit || ProbotOctokit,
      appId: opts.id,
      privateKey: opts.privateKey,
      cache,
    });

    this.state = {
      cache,
      githubToken: opts.githubToken,
      log: this.log,
      Octokit,
      octokit: opts.octokit || new Octokit(),
      throttleOptions: getThrottleOptions({
        log: this.log,
        throttleOptions: opts.throttleOptions,
        redisConfig: options.redisConfig,
      }),
    };

    this.router = opts.router || express.Router(); // you can do this?

    this.webhooks =
      opts.webhooks ||
      new Webhooks<Context>({
        path: options.webhookPath,
        secret: options.secret,
        transform: webhookTransform.bind(null, this.state),
      });

    this.on = (eventNameOrNames, callback) => {
      // when an app subscribes to an event using `app.on(event, callback)`, Probot sends a request to `GET /app` and
      // verifies if the app is subscribed to the event and logs a warning if it is not.
      //
      // This feature will be moved out of Probot core as it has side effects and does not work in a stateless environment.
      webhookEventCheck(this.state, eventNameOrNames);

      return this.webhooks.on(eventNameOrNames, callback);
    };
    this.receive = this.webhooks.receive;

    // TODO: do not add error handler if `opts.webhooks` was passed
    this.webhooks.on("error", (error) => {
      // avoid the error.code deprecation message
      // can be replaced with `log.error({ err, event, ...err })` once @octokit/request-error v3 is used

      const { name, message, stack, event } = error;

      if (!event) {
        const log = this.log.child({ name: "event" });
        log.error({
          err: {
            name,
            message,
            stack,
          },
        });
        return;
      }

      for (const { headers, request, status } of error) {
        const log = this.log.child({ name: "event", id: event.id });

        log.error({
          err: {
            name,
            message,
            stack,
          },
          event,
          headers,
          request,
          status,
        });
      }
    });
  }

  /**
   * Loads an ApplicationFunction into the current Application
   * @param appFn - Probot application function to load
   */
  public load(appFn: ApplicationFunction | ApplicationFunction[]): Application {
    if (Array.isArray(appFn)) {
      appFn.forEach((a) => this.load(a));
    } else {
      appFn(this);
    }

    return this;
  }

  /**
   * Get an {@link http://expressjs.com|express} router that can be used to
   * expose HTTP endpoints
   *
   * ```
   * module.exports = app => {
   *   // Get an express router to expose new HTTP endpoints
   *   const route = app.route('/my-app');
   *
   *   // Use any middleware
   *   route.use(require('express').static(__dirname + '/public'));
   *
   *   // Add a new route
   *   route.get('/hello-world', (req, res) => {
   *     res.end('Hello World');
   *   });
   * };
   * ```
   *
   * @param path - the prefix for the routes
   * @returns an [express.Router](http://expressjs.com/en/4x/api.html#router)
   */
  public route(path?: string): express.Router {
    if (path) {
      const router = express.Router();
      this.router.use(path, router);
      return router;
    } else {
      return this.router;
    }
  }

  /**
   * Authenticate and get a GitHub client that can be used to make API calls.
   *
   * You'll probably want to use `context.github` instead.
   *
   * **Note**: `app.auth` is asynchronous, so it needs to be prefixed with a
   * [`await`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await)
   * to wait for the magic to happen.
   *
   * ```js
   *  module.exports = (app) => {
   *    app.on('issues.opened', async context => {
   *      const github = await app.auth();
   *    });
   *  };
   * ```
   *
   * @param id - ID of the installation, which can be extracted from
   * `context.payload.installation.id`. If called without this parameter, the
   * client wil authenticate [as the app](https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/about-authentication-options-for-github-apps/#authenticating-as-a-github-app)
   * instead of as a specific installation, which means it can only be used for
   * [app APIs](https://developer.github.com/v3/apps/).
   *
   * @returns An authenticated GitHub API client
   */
  public async auth(
    installationId?: number,
    log?: LoggerWithTarget
  ): Promise<InstanceType<typeof ProbotOctokit>> {
    return getAuthenticatedOctokit(
      Object.assign({}, this.state, log ? { log } : null),
      installationId
    );
  }
}
