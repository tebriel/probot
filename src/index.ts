// tslint:disable-next-line: no-var-requires
require("dotenv").config();

import { WebhookEvent, Webhooks } from "@octokit/webhooks";
import express from "express";
import Redis from "ioredis";
import LRUCache from "lru-cache";
import { Deprecation } from "deprecation";
import pinoHttp from "pino-http";

import type { Logger } from "pino";

import { Server } from "http";
import { Application } from "./application";
import { setupApp } from "./apps/setup";
import { Context } from "./context";
import { ProbotOctokit } from "./github/octokit";
import { getLog } from "./get-log";
import { findPrivateKey } from "./private-key";
import { resolve } from "./resolver";
import { createServer } from "./server";
import { createWebhookProxy } from "./webhook-proxy";
import { getErrorHandler } from "./error-handler";
import { ProbotWebhooks, State } from "./types";
import { webhookTransform } from "./webhook-transform";
import { getThrottleOptions } from "./get-throttle-options";
import { getProbotOctokitWithDefaults } from "./get-probot-octokit-with-defaults";

import { handleDeprecatedEnvironmentVariables } from "./handle-deprecated-environment-variables";
handleDeprecatedEnvironmentVariables();

export interface Options {
  // same options as Application class
  privateKey?: string;
  githubToken?: string;
  id?: number;
  Octokit?: typeof ProbotOctokit;
  log?: Logger;
  redisConfig?: Redis.RedisOptions;
  secret?: string;
  webhookPath?: string;

  // Probot class-specific options
  /**
   * @deprecated `cert` options is deprecated. Use `privateKey` instead
   */
  cert?: string;
  port?: number;
  webhookProxy?: string;
}

// tslint:disable:no-var-requires
const defaultAppFns: ApplicationFunction[] = [require("./apps/default")];
// tslint:enable:no-var-requires

export class Probot {
  public static async run(appFn: ApplicationFunction | string[]) {
    const pkgConf = require("pkg-conf");
    const program = require("commander");

    const readOptions = (): Options => {
      if (Array.isArray(appFn)) {
        program
          .usage("[options] <apps...>")
          .option(
            "-p, --port <n>",
            "Port to start the server on",
            process.env.PORT || 3000
          )
          .option(
            "-W, --webhook-proxy <url>",
            "URL of the webhook proxy service.`",
            process.env.WEBHOOK_PROXY_URL
          )
          .option(
            "-w, --webhook-path <path>",
            "URL path which receives webhooks. Ex: `/webhook`",
            process.env.WEBHOOK_PATH
          )
          .option("-a, --app <id>", "ID of the GitHub App", process.env.APP_ID)
          .option(
            "-s, --secret <secret>",
            "Webhook secret of the GitHub App",
            process.env.WEBHOOK_SECRET
          )
          .option(
            "-P, --private-key <file>",
            "Path to certificate of the GitHub App",
            process.env.PRIVATE_KEY_PATH
          )
          .parse(appFn);

        return {
          privateKey: findPrivateKey(program.privateKey) || undefined,
          id: program.app,
          port: program.port,
          secret: program.secret,
          webhookPath: program.webhookPath,
          webhookProxy: program.webhookProxy,
        };
      }
      const privateKey = findPrivateKey();
      return {
        privateKey: (privateKey && privateKey.toString()) || undefined,
        id: Number(process.env.APP_ID),
        port: Number(process.env.PORT) || 3000,
        secret: process.env.WEBHOOK_SECRET,
        webhookPath: process.env.WEBHOOK_PATH,
        webhookProxy: process.env.WEBHOOK_PROXY_URL,
      };
    };

    const options = readOptions();
    const probot = new Probot(options);

    if (!options.id || !options.privateKey) {
      if (process.env.NODE_ENV === "production") {
        if (!options.id) {
          throw new Error(
            "Application ID is missing, and is required to run in production mode. " +
              "To resolve, ensure the APP_ID environment variable is set."
          );
        } else if (!options.privateKey) {
          throw new Error(
            "Certificate is missing, and is required to run in production mode. " +
              "To resolve, ensure either the PRIVATE_KEY or PRIVATE_KEY_PATH environment variable is set and contains a valid certificate"
          );
        }
      }
      probot.load(setupApp);
    } else if (Array.isArray(appFn)) {
      const pkg = await pkgConf("probot");
      probot.setup(program.args.concat(pkg.apps || pkg.plugins || []));
    } else {
      probot.load(appFn);
    }
    probot.start();

    return probot;
  }

  public server: express.Application;
  public webhooks: ProbotWebhooks;
  public log: Logger;

  // These 3 need to be public for the tests to work.
  public options: Options;
  public throttleOptions: any;

  private httpServer?: Server;
  private apps: Application[];
  private state: State;

  /**
   * @deprecated use probot.log instead
   */
  public get logger() {
    this.log.warn(
      new Deprecation(
        `[probot] "probot.logger" is deprecated. Use "probot.log" instead`
      )
    );
    return this.log;
  }

  constructor(options: Options) {
    if (process.env.GHE_HOST && /^https?:\/\//.test(process.env.GHE_HOST)) {
      throw new Error(
        "Your `GHE_HOST` environment variable should not begin with https:// or http://"
      );
    }

    //
    // Probot class-specific options (Express server & Webhooks)
    //
    options.webhookPath = options.webhookPath || "/";
    options.secret = options.secret || "development";

    // TODO: deprecate probot.log in favor of probot.log.
    this.log = options.log || getLog();

    if (options.cert) {
      this.log.warn(
        new Deprecation(
          `[probot] "cert" option is deprecated. Use "privateKey" instead`
        )
      );
      options.privateKey = options.cert;
    }

    this.apps = [];

    // TODO: Why is this public?
    this.options = options;

    // TODO: support redis backend for access token cache if `options.redisConfig || process.env.REDIS_URL`
    const cache = new LRUCache<number, string>({
      // cache max. 15000 tokens, that will use less than 10mb memory
      max: 15000,
      // Cache for 1 minute less than GitHub expiry
      maxAge: Number(process.env.INSTALLATION_TOKEN_TTL) || 1000 * 60 * 59,
    });

    const Octokit = getProbotOctokitWithDefaults({
      githubToken: options.githubToken,
      Octokit: options.Octokit || ProbotOctokit,
      appId: options.id,
      privateKey: options.privateKey,
      cache,
    });
    const octokit = new Octokit();

    this.throttleOptions = getThrottleOptions({
      log: this.log,
      redisConfig: options.redisConfig,
    });

    this.state = {
      cache,
      githubToken: options.githubToken,
      log: this.log,
      Octokit,
      octokit,
      throttleOptions: this.throttleOptions,
    };

    this.webhooks = new Webhooks({
      path: options.webhookPath,
      secret: options.secret,
      transform: webhookTransform.bind(null, this.state),
    });
    // TODO: normalize error handler with code in application.ts
    this.webhooks.on("error", getErrorHandler(this.log));

    this.server = createServer({
      webhook: (this.webhooks as any).middleware,
      logger: this.log,
    });
  }

  /**
   * @deprecated `probot.webhook` is deprecated. Use `probot.webhooks` instead
   */
  public get webhook(): Webhooks {
    this.log.warn(
      new Deprecation(
        `[probot] "probot.webhook" is deprecated. Use "probot.webhooks" instead instead`
      )
    );

    return this.webhooks;
  }

  public receive(event: WebhookEvent) {
    this.log.debug({ event }, "Webhook received");
    return Promise.all(this.apps.map((app) => app.receive(event)));
  }

  public load(appFn: string | ApplicationFunction) {
    if (typeof appFn === "string") {
      appFn = resolve(appFn) as ApplicationFunction;
    }

    const app = new Application({
      cache: this.state.cache,
      githubToken: this.state.githubToken,
      Octokit: this.state.Octokit,
      octokit: this.state.octokit,
      throttleOptions: this.throttleOptions,
      webhooks: this.webhooks,
    });

    // Connect the router from the app to the server
    this.server.use(app.router);

    // Initialize the ApplicationFunction
    app.load(appFn);
    this.apps.push(app);

    return app;
  }

  public setup(appFns: Array<string | ApplicationFunction>) {
    // Log all unhandled rejections
    (process as NodeJS.EventEmitter).on("unhandledRejection", getErrorHandler);

    // Load the given appFns along with the default ones
    appFns.concat(defaultAppFns).forEach((appFn) => this.load(appFn));

    // Register error handler as the last middleware
    this.server.use(
      pinoHttp({
        logger: this.log,
      })
    );
  }

  public start() {
    this.httpServer = this.server
      .listen(this.options.port, () => {
        if (this.options.webhookProxy) {
          createWebhookProxy({
            logger: this.log,
            path: this.options.webhookPath,
            port: this.options.port,
            url: this.options.webhookProxy,
          });
        }
        this.log.info("Listening on http://localhost:" + this.options.port);
      })
      .on("error", (err: NodeJS.ErrnoException) => {
        if (err.code === "EADDRINUSE") {
          this.log.error(
            `Port ${this.options.port} is already in use. You can define the PORT environment variable to use a different port.`
          );
        } else {
          this.log.error(err);
        }
        process.exit(1);
      });

    return this.httpServer;
  }

  public stop() {
    if (!this.httpServer) return;

    this.httpServer.close();
  }
}

export const createProbot = (options: Options) => {
  options.log = options.log || getLog();
  options.log.warn(
    new Deprecation(
      `[probot] "createProbot(options)" is deprecated, use "new Probot(options)" instead`
    )
  );
  return new Probot(options);
};

export type ApplicationFunction = (app: Application) => void;

export { Logger, Context, Application, ProbotOctokit };
