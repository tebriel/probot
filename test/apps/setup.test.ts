const createChannel = jest.fn().mockResolvedValue("mocked proxy URL");
const updateDotenv = jest.fn().mockResolvedValue({});
jest.mock("smee-client", () => ({ createChannel }));
jest.mock("update-dotenv", () => updateDotenv);

import nock from "nock";
import request from "supertest";
import { Probot } from "../../src";
import { setupApp } from "../../src/apps/setup";

describe("Setup app", () => {
  let probot: Probot;

  beforeEach(async () => {
    delete process.env.WEBHOOK_PROXY_URL;
    probot = new Probot({});
    probot.load(setupApp);

    // there is currently no way to await probot.load, so we do hacky hack hack
    await new Promise((resolve) => setTimeout(resolve, 100));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("GET /probot", () => {
    it("returns a 200 response", async () => {
      await request(probot.server).get("/probot").expect(200);
    });
  });

  describe("GET /probot/setup", () => {
    it("returns a redirect", async () => {
      nock("https://api.github.com")
        .post("/app-manifests/123/conversions")
        .reply(201, {
          html_url: "/apps/my-app",
          id: "id",
          pem: "pem",
          webhook_secret: "webhook_secret",
        });

      await request(probot.server)
        .get("/probot/setup")
        .query({ code: "123" })
        .expect(302)
        .expect("Location", "/apps/my-app/installations/new");

      expect(createChannel).toHaveBeenCalledTimes(1);
      expect(updateDotenv.mock.calls).toMatchSnapshot();
    });
  });

  describe("GET /probot/success", () => {
    it("returns a 200 response", async () => {
      await request(probot.server).get("/probot/success").expect(200);

      expect(createChannel).toHaveBeenCalledTimes(1);
    });
  });
});
