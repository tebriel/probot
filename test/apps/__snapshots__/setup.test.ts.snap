// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Setup app GET /probot/setup returns a redirect 1`] = `
Array [
  Array [
    Object {
      "WEBHOOK_PROXY_URL": "mocked proxy URL",
    },
  ],
  Array [
    Object {
      "APP_ID": "id",
      "PRIVATE_KEY": "\\"pem\\"",
      "WEBHOOK_SECRET": "webhook_secret",
    },
  ],
]
`;
